### ISSUE TITLE
Profile-profile database clustering

### ISSUE THREAD

**Issue by ksteczk** [soedinglab/MMseqs2#66]
I know that could be quite unpopular clustering strategy, but did you consider clustering based on comparing profiles? That would require adding additional step in cascade clustering for clustering within PSI-BLAST distances - I wondered if you have worked out some strategy on that. I'm going to cluster the whole nr database in that way so I'm exploring the possibilities. I expect that expanding -e option to, let it be 0.05 won't give results comparable to evalue's of e.g. 0.005 but with PSI-BLAST (or profile searches with MMseqs2, which I'm going to use).

**Comment by martin-steinegger** [comment]
Sorry for the late answer @ksteczk. 
What do you mean by PSI-BLAST distances? 
To cluster with profiles, you could first build profiles:

     mmseqs search NR NR NR_res /tmp --num-iterations 3 
     mmseqs result2profile NR NR NR_res NR_profile

And then cluster them:

     mmseqs search NR_profile NR_consensus NR_pp_res /tmp  # Add your cluster criteria here
     mmseqs clust NR_profile NR_pp_res NR_pp_clu

**Comment by ksteczk** [comment]
Thank you for the workflow. I'll try that. For now as I explore the software a little bit (I'm amazed with its speed and scalability) I cluster smaller database.

A part of my workflow is generating profiles for pdb_05_rep database based on nr70_rep database:
```
mmseqs search pdb_05_rep nr70_rep pdb_05_res tmp --num-iterations 3 
mmseqs result2profile pdb_05_rep nr70_rep pdb_05_res pdb_05_profile --threads 110

```
The second command however blasts my bash with these warnings:

> Warning in /opt/install/MMseq2/src/alignment/MsaFilter.cpp:244: filter:
>         Filtering removed all sequences in alignment. Inserting back first sequence.

Is that OK? Or something went wrong with the procedure?

**Comment by martin-steinegger** [comment]
I assume the message comes from alignments that find --max-seqs similiar sequences. So the filter would remove all sequences but then add back the query sequence. I would recommend to work on a redundancy reduced database. E.g. Uniclust90  (uniclust.mmseqs.com).

I forgot, you need to add the flat --add-self-matches to your search. 

     mmseqs search pdb_05_profile pdb_05_rep pdb_05_pp_res tmp  --threads 110 -e 0.05 --add-self-matches
     mmseqs clust pdb_05_profile pdb_05_pp_res pdb_05_pp_clu  --threads 110 --cluster-mode 1 --similarity-type 1

Do you use MMseqs2 on a computer with 110 core?

**Comment by ksteczk** [comment]
Thanks for the hint. I'll try that.

As to computer - yes, I'm using 120 cores single machine. The software
scales up amazingly. Now I'm computing 6it profiles for 31M sequences and
that will be accomplished in 3 days.

In few days I'll start tests also with MPI run, so that mmseqs2 will be ran
on 1600 CPUs. I'll give you a feedback on the speed soon then.


czw., 1.03.2018, 15:52 użytkownik Martin Steinegger <
notifications@github.com> napisał:

**Comment by martin-steinegger** [comment]
Thanks for your detailed report! The issue you are pointing out is indeed a barrier. I noticed this myself.  We remap the mmaped prefilter result file every 1 million queries. 

I did this since I noticed a huge slow down if the memory of the machine is not enough to hold the whole alignment file in memory. It seems that the kernel paging is quite slow. 

I maybe should add a check if the machine has enough memory to keep the whole file in memory.

**Comment by ksteczk** [comment]
> I maybe should add a check if the machine has enough memory to keep the whole file in memory.

That would be awesome! - I have a machine with 6TB of RAM where I run this clustering so it could handle all of that in memory.

**Comment by milot-mirdita** [comment]
Martin's fix in db458278e0a42beefe93cb2df4eebfd13cf67d1c should fix the slowdown issue on your server.

**Comment by martin-steinegger** [comment]
I close this issue now. Please reopen if you still have issues with the profile-profile clustering.
